File: Data Structures & Algorithms/max-water-container/submission-11.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        # 2 ptr approach
        res = 0
        l, r = 0, len(heights) - 1

        while l < r:
            currentArea = min(heights[l], heights[r]) * (r - l)
            res = max(res, currentArea)

            if heights[l] < heights[r]:
                l += 1
            else:
                r -= 1
        
        return res
   

File: Data Structures & Algorithms/trapping-rain-water/submission-11.java
class Solution {
    public int trap(int[] height) {
        if (height.length == 0) {
            return 0;
        }

        int l = 0, r = height.length - 1;
        int leftMax = height[l], rightMax = height[r];
        int res = 0;

        while (l < r) {
            // the left is the bottleneck
            if (leftMax < rightMax) {
                res += Math.min(leftMax, rightMax) - height[l];
                l++;
                leftMax = Math.max(leftMax, height[l]);
            
            // right is the bottleneck
            } else {
                res += Math.min(leftMax, rightMax) - height[r];
                r--;
                rightMax = Math.max(rightMax, height[r]);
            }
        }
        return res;
    }
}
